### ISSUE TITLE
For example code of "Interoperability with bilinear attention visualization".

### ISSUE THREAD

**Issue by DeepMindAGI** [pz-white/DrugBAN#2]
Hello, I am very interested in your research. Do you have an example code of "Interoperability with linear attention visualization"？

**Comment by peizhenbai** [comment]
Hi,

Thanks for your interest in our work. The ligand-protein interaction maps and 3D crystal structures are visualized using MOE software. The 2D structures with highlighted atoms are visualized using RDKit package. We first calculate top 20% weighted atoms from the bilinear attention map, and then highlight them by RDKit. You can refer to the documentation: http://rdkit.blogspot.com/2020/01/similarity-maps-with-new-drawing-code.html. Here is a visualization example:
```
colorMap = LinearSegmentedColormap.from_list('Wistia', [cmap(0), (1.0, 1.0, 1.0), cmap(1)], N=255)
mol = Chem.MolFromSmiles(smiles)
SimilarityMaps.GetSimilarityMapFromWeights(mol, atom_weights, draw2d=d, contourLines=0, colorMap=colorMap)
d.FinishDrawing()
show_png(mol_d.GetDrawingText())
```

**Comment by DeepMindAGI** [comment]
> Hi,
> 
> Thanks for your interest in our work. The ligand-protein interaction maps and 3D crystal structures are visualized using MOE software. The 2D structures with highlighted atoms are visualized using RDKit package. We first calculate top 20% weighted atoms from the bilinear attention map, and then highlight them by RDKit. You can refer to the documentation: http://rdkit.blogspot.com/2020/01/similarity-maps-with-new-drawing-code.html. Here is a visualization example:
> 
> ```
> colorMap = LinearSegmentedColormap.from_list('Wistia', [cmap(0), (1.0, 1.0, 1.0), cmap(1)], N=255)
> mol = Chem.MolFromSmiles(smiles)
> SimilarityMaps.GetSimilarityMapFromWeights(mol, atom_weights, draw2d=d, contourLines=0, colorMap=colorMap)
> d.FinishDrawing()
> show_png(mol_d.GetDrawingText())
> ```

Thank you. Can you also provide the sample code about using the training model to predict? Thank you very much! 感谢大佬哈哈

**Comment by peizhenbai** [comment]
Hi,

The model prediction can be obtained by the model module after training. You just need normalize the bilinear attention map and sort the top weights for atoms/residues, respectively. I do not keep the sample code and hope this can help you.

**Comment by peizhenbai** [comment]
Hi, the model prediction can be obtained by the model module after training. You just need normalize the bilinear attention map and sort the top weights for atoms/residues, respectively. I do not keep the sample code and hope this can help you.

**Comment by zpking** [comment]
`class DrugBAN(nn.Module):
    def __init__(self, **config):
        super(DrugBAN, self).__init__()
        drug_in_feats = config["DRUG"]["NODE_IN_FEATS"]
        drug_embedding = config["DRUG"]["NODE_IN_EMBEDDING"]
        drug_hidden_feats = config["DRUG"]["HIDDEN_LAYERS"]
        protein_emb_dim = config["PROTEIN"]["EMBEDDING_DIM"]
        num_filters = config["PROTEIN"]["NUM_FILTERS"]
        kernel_size = config["PROTEIN"]["KERNEL_SIZE"]
        mlp_in_dim = config["DECODER"]["IN_DIM"]
        mlp_hidden_dim = config["DECODER"]["HIDDEN_DIM"]
        mlp_out_dim = config["DECODER"]["OUT_DIM"]
        drug_padding = config["DRUG"]["PADDING"]
        protein_padding = config["PROTEIN"]["PADDING"]
        out_binary = config["DECODER"]["BINARY"]
        ban_heads = config["BCN"]["HEADS"]
        self.drug_extractor = MolecularGCN(in_feats=drug_in_feats, dim_embedding=drug_embedding,
                                           padding=drug_padding,
                                           hidden_feats=drug_hidden_feats)
        self.protein_extractor = ProteinCNN(protein_emb_dim, num_filters, kernel_size, protein_padding)

        self.bcn = weight_norm(
            BANLayer(v_dim=drug_hidden_feats[-1], q_dim=num_filters[-1], h_dim=mlp_in_dim, h_out=ban_heads),
            name='h_mat', dim=None)
        self.mlp_classifier = MLPDecoder(mlp_in_dim, mlp_hidden_dim, mlp_out_dim, binary=out_binary)

    def forward(self, bg_d, v_p, mode="train"):
        v_d = self.drug_extractor(bg_d)
        v_p = self.protein_extractor(v_p)
        f, att = self.bcn(v_d, v_p)
        score = self.mlp_classifier(f)
        if mode == "train":
            return v_d, v_p, f, score
        elif mode == "eval":
            return v_d, v_p, score, att
`
Thank you for your answer, there's another question
Is the return value _att_ in this model the attention weight you mean?
**Torch. Size ([1,2,2901185])**, where 290 is an atom and 1185 is a residue. Am I correct in my understanding?

**Comment by zpking** [comment]
`class DrugBAN(nn.Module):
    def __init__(self, **config):
        super(DrugBAN, self).__init__()
        drug_in_feats = config["DRUG"]["NODE_IN_FEATS"]
        drug_embedding = config["DRUG"]["NODE_IN_EMBEDDING"]
        drug_hidden_feats = config["DRUG"]["HIDDEN_LAYERS"]
        protein_emb_dim = config["PROTEIN"]["EMBEDDING_DIM"]
        num_filters = config["PROTEIN"]["NUM_FILTERS"]
        kernel_size = config["PROTEIN"]["KERNEL_SIZE"]
        mlp_in_dim = config["DECODER"]["IN_DIM"]
        mlp_hidden_dim = config["DECODER"]["HIDDEN_DIM"]
        mlp_out_dim = config["DECODER"]["OUT_DIM"]
        drug_padding = config["DRUG"]["PADDING"]
        protein_padding = config["PROTEIN"]["PADDING"]
        out_binary = config["DECODER"]["BINARY"]
        ban_heads = config["BCN"]["HEADS"]
        self.drug_extractor = MolecularGCN(in_feats=drug_in_feats, dim_embedding=drug_embedding,
                                           padding=drug_padding,
                                           hidden_feats=drug_hidden_feats)
        self.protein_extractor = ProteinCNN(protein_emb_dim, num_filters, kernel_size, protein_padding)

        self.bcn = weight_norm(
            BANLayer(v_dim=drug_hidden_feats[-1], q_dim=num_filters[-1], h_dim=mlp_in_dim, h_out=ban_heads),
            name='h_mat', dim=None)
        self.mlp_classifier = MLPDecoder(mlp_in_dim, mlp_hidden_dim, mlp_out_dim, binary=out_binary)

    def forward(self, bg_d, v_p, mode="train"):
        v_d = self.drug_extractor(bg_d)
        v_p = self.protein_extractor(v_p)
        f, att = self.bcn(v_d, v_p)
        score = self.mlp_classifier(f)
        if mode == "train":
            return v_d, v_p, f, score
        elif mode == "eval":
            return v_d, v_p, score, att`

Thank you for your answer, there's another question
Is the return value _att_ in this model the attention weight you mean?
**Torch. Size ([1,2,2901185])**, where 290 is an atom and 1185 is a residue. Am I correct in my understanding?
